File: scripts/algorithms/O/Odd Even Linked List/Odd Even Linked List.java
// Runtime: 0 ms (Top 100.00%) | Memory: 44.6 MB (Top 61.89%)
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if(head == null) {
            return head;
        }
        ListNode result = head, evenHalf = new ListNode(0), evenHalfPtr = evenHalf;
        for(; head.next != null; head = head.next) {
            evenHalfPtr = evenHalfPtr.next = head.next;
            head.next = head.next.next;
            evenHalfPtr.next = null;
            if(head.next == null) {
                break;
            }
        }
        head.next = evenHalf.next;
        return result;
    }
}

File: scripts/algorithms/O/Odd Even Linked List/Odd Even Linked List.cpp
// Runtime: 19 ms (Top 54.79%) | Memory: 10.4 MB (Top 74.28%)
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == NULL || head->next == NULL || head->next->next == NULL){
            return head;
        }
        ListNode *first = head;
        ListNode *second = head->next;
        ListNode *last = head;
        int count = 1;
        while(last->next != NULL){
            last = last->next;
            count++;
        }

        int n = count/2;
        while(n != 0){
            first->next = second->next;
            last->next = second;
            second->next = NULL;
            first = first->next;
            last = second;
            second = first->next;
            n--;
        }
        return head;
    }
};